### PR TITLE
Add Machine Learning filter APIs and detector custom rules

### PR DESCRIPTION
**Pull request by russcam**
Relates: #3615

This PR adds 

- the Put, Get, Update and Delete filters APIs, required for working with ML detector custom rules
- custom rules to machine learning detectors. The name of types used closely follows those used in Elasticsearch source.

### CODE REVIEW

File: src/CodeGeneration/ApiGenerator/Domain/ApiQueryParameters.cs [hunk 1]
```diff
@@ -104,6 +104,7 @@ public string TypeLowLevel
 				{
 					case "boolean": return "bool?";
 					case "list": return "string[]";
+					case "int":
```
Comment (codebrain) [inline src/CodeGeneration/ApiGenerator/Domain/ApiQueryParameters.cs]: I'm wondering if we ought to raise an issue for this to change from `int` to `integer`?

Comment (russcam) [reply src/CodeGeneration/ApiGenerator/Domain/ApiQueryParameters.cs]: issue or pull request 👍

File: src/Nest/XPack/MachineLearning/GetFilters/GetFiltersResponse.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		/// The count of filters.
+		/// </summary>
+		[JsonProperty("count")]
+		long Count { get; }
+
+		/// <summary>
+		/// An array of filters resources
+		/// </summary>
+		[JsonProperty("filters")]
+		IReadOnlyCollection<Filter> Filters { get; }
```
Comment (codebrain) [inline src/Nest/XPack/MachineLearning/GetFilters/GetFiltersResponse.cs]: Init empty collection?

File: src/Nest/XPack/MachineLearning/PutFilter/PutFilterRequest.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace Nest
+{
+	/// <summary>
+	/// Creates a machine learning filter.
+	/// </summary>
+	public partial interface IPutFilterRequest
+	{
+		[JsonProperty("description")]
+		string Description { get; set; }
```
Comment (codebrain) [inline src/Nest/XPack/MachineLearning/PutFilter/PutFilterRequest.cs]: Docs?

File: src/Nest/XPack/MachineLearning/GetFilters/GetFiltersResponse.cs (hunk 2 above)
Comment (russcam) [reply src/Nest/XPack/MachineLearning/GetFilters/GetFiltersResponse.cs]: It's done on `GetFiltersResponse`

File: src/Nest/XPack/MachineLearning/PutFilter/PutFilterRequest.cs (hunk 3 above)
Comment (russcam) [reply src/Nest/XPack/MachineLearning/PutFilter/PutFilterRequest.cs]: Added 👍

File: src/CodeGeneration/ApiGenerator/Domain/ApiQueryParameters.cs (hunk 1 above)
Comment (russcam) [reply src/CodeGeneration/ApiGenerator/Domain/ApiQueryParameters.cs]: Looking at the specs, it looks like they no longer contain `"integer"`, so I'll remove that case

**Pull request merged**